micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jpa/HibernateQueryMetrics.java

```diff
@@ -0,0 +1,181 @@
+
+        MetricsEventHandler(MeterRegistry meterRegistry) {
+            this.meterRegistry = meterRegistry;
+        }
+
+        @Override
+        public void onPostLoad(PostLoadEvent event) {
+            registerQueryMetric(event.getSession().getFactory().getStatistics());
+        }
+
+        void registerQueryMetric(Statistics statistics) {
+            for (String query : statistics.getQueries()) {
+                QueryStatistics queryStatistics = statistics.getQueryStatistics(query);
+                if (Objects.nonNull(queryStatistics)) {
+                    String queryName = query.replace(" ", "_").replace(".", "_").toLowerCase();
+                    if (Objects.isNull(Search.in(meterRegistry).tags("query", queryName).functionCounter())) {
+
+                        FunctionCounter.builder("hibernate.query.cache.requests", queryStatistics, QueryStatistics::getCacheHitCount)
+                                .tags(tags)
+                                .tags("result", "hit", "query", queryName)
+                                .description("The number of query cache hits")
+                                .register(meterRegistry);
+
+                        FunctionCounter.builder("hibernate.query.cache.requests", queryStatistics, QueryStatistics::getCacheMissCount)
+                                .tags(tags)
+                                .tags("result", "miss", "query", queryName)
+                                .description("The number of query cache miss")
+                                .register(meterRegistry);
+
+                        FunctionCounter.builder("hibernate.query.cache.puts", queryStatistics, QueryStatistics::getCacheMissCount)
+                                .tags(tags)
+                                .tags("query", queryName)
+                                .description("The number of putting the query into cache")
+                                .register(meterRegistry);
+
+                        FunctionTimer.builder("hibernate.query.executions.total", queryStatistics, QueryStatistics::getExecutionCount, QueryStatistics::getExecutionTotalTime, TimeUnit.MILLISECONDS)
+                                .tags(tags)
+                                .tags("query", queryName)
+                                .description("Function tracked total number of query executions during time")
+                                .register(meterRegistry);
+
+                        TimeGauge.builder("hibernate.query.executions.avg", queryStatistics, TimeUnit.MILLISECONDS, QueryStatistics::getExecutionAvgTime)
+                                .tags(tags)
+                                .tags("query", queryName)
+                                .description("Query average execution time")
+                                .register(meterRegistry);
+
+                        TimeGauge.builder("hibernate.query.executions.max", queryStatistics, TimeUnit.MILLISECONDS, QueryStatistics::getExecutionMaxTime)
```

Comment:
This will clash with the metric by the same name in `HibernateMetrics`: https://github.com/micrometer-metrics/micrometer/blob/516f9c51bd6f4ed7e7b418a98c04a2d5c432f940/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/jpa/HibernateMetrics.java#L216-L219

If we have per query metrics, we don't need the overall metric from `HibernateMetrics`, but we won't know if both binders are going to be used or not.

Reply:
Won't this clash be resolved by query tag?
The general metrics for hibernate will be put under:
hibernate/query/executions/max
The query metrics will be put under:
hibernate/query/executions/max/QUERY_TAG

Reply:
Having the same metric name with a different set of tag names is not allowed with the Prometheus backend, at least. I need to double check if that restriction is still necessary as there was a suggestion it may no longer be.

Reply:
Okey, let's resolve this clash then. I've changed .executions -> .execution which kinda make sense since it is a single query.